components/brave_wallet/browser/zcash/rust/orchard_shard_tree_impl.cc

```diff
@@ -0,0 +1,358 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/zcash/rust/orchard_shard_tree_impl.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "base/memory/ptr_util.h"
+#include "base/ranges/algorithm.h"
+#include "brave/components/brave_wallet/browser/zcash/rust/cxx/src/shard_store.h"
+#include "brave/components/brave_wallet/browser/zcash/rust/orchard_decoded_blocks_bunde_impl.h"
+#include "brave/components/brave_wallet/common/hex_utils.h"
+#include "brave/components/brave_wallet/common/zcash_utils.h"
+
+namespace brave_wallet::orchard {
+
+::brave_wallet::OrchardShardAddress From(const ShardTreeAddress& addr) {
+  return ::brave_wallet::OrchardShardAddress{addr.level, addr.index};
+}
+
+ShardTreeAddress From(const ::brave_wallet::OrchardShardAddress& addr) {
+  return ShardTreeAddress{addr.level, addr.index};
+}
+
+ShardTreeCap From(::brave_wallet::OrchardShardTreeCap& shard_store_cap) {
+  ::rust::Vec<uint8_t> data;
+  data.reserve(shard_store_cap.size());
+  base::ranges::copy(shard_store_cap, std::back_inserter(data));
+  return ShardTreeCap{std::move(data)};
+}
+
+::brave_wallet::OrchardShardTreeCap From(const ShardTreeCap& cap) {
+  ::brave_wallet::OrchardShardTreeCap shard_store_cap;
+  shard_store_cap.reserve(cap.data.size());
+  base::ranges::copy(cap.data, std::back_inserter(shard_store_cap));
+  return shard_store_cap;
```

Comment:
It makes sense to use/add some utility functions so conversions between rust::cxxbridge1::Vec<T> and std::vector<T> could be written as oneliners